src/repr/src/fixed_length.rs

```diff
@@ -0,0 +1,752 @@
+/// is reserved for nullability encoding.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct BytesN<const N: usize> {
+    #[serde(with = "BigArray")]
+    data: [u8; N],
+}
+
+// TODO(vmarcos): We may be able to simplify the implementation of Columnation
+// once frankmcsherry/columnation#18 lands.
+impl<const N: usize> Columnation for BytesN<N> {
+    type InnerRegion = CopyRegion<BytesN<N>>;
+}
+
+/// Helper function to read 8 bits as `bool`s out of a `u8`.
+fn u8_to_bool_slice(b: u8) -> [bool; 8] {
+    let mut result = [false; 8];
+    for i in 0..8 {
+        result[i] = (b & (1u8 << i)) != 0u8;
+    }
+    result
+}
+
+/// Helper function to obtain the fixed length of a `ScalarType`, if there is one.
+pub(super) fn fixed_length(typ: &ScalarType) -> Option<usize> {
+    match typ {
+        ScalarType::Bool => Some(size_of::<u8>()),
+        ScalarType::Int16 => Some(size_of::<i16>()),
+        ScalarType::Int32 => Some(size_of::<i32>()),
+        ScalarType::Int64 => Some(size_of::<i64>()),
+        ScalarType::UInt16 => Some(size_of::<u16>()),
+        ScalarType::UInt32 => Some(size_of::<u32>()),
+        ScalarType::UInt64 => Some(size_of::<u64>()),
+        ScalarType::Float32 => Some(size_of::<f32>()),
+        ScalarType::Float64 => Some(size_of::<f64>()),
+        ScalarType::Date => Some(size_of::<i32>()),
+        ScalarType::Time => Some(2 * size_of::<u32>()),
+        ScalarType::Timestamp { precision: _ } => Some(3 * size_of::<i32>() + size_of::<u32>()),
+        ScalarType::TimestampTz { precision: _ } => Some(3 * size_of::<i32>() + size_of::<u32>()),
+        ScalarType::Interval => Some(2 * size_of::<i32>() + size_of::<i64>()),
+        ScalarType::Uuid => Some(size_of::<uuid::Bytes>()),
+        ScalarType::Oid => Some(size_of::<u32>()),
+        ScalarType::MzTimestamp => Some(size_of::<Timestamp>()),
+        ScalarType::MzAclItem => Some(MzAclItem::binary_size()),
+        ScalarType::AclItem => Some(AclItem::binary_size()),
+        _ => None,
+    }
+}
+
+impl<'a, const N: usize> IntoRowByTypes<'a> for BytesN<N> {
+    /// This implementation employs the schema provided by `types` to decode
+    /// the N-byte value into `row_buf`, returning a reference to it.
+    ///
+    /// Presently, the implementation panics if more than one `ColumnType` is found
+    /// in `types` or if the one column has a byte length greater than N-1 (due to one
+    /// byte being reserved for nullability encoding).
+    fn into_row(&'a self, row_buf: &'a mut Row, types: &[ColumnType]) -> &'a Row {
+        // TODO (vmarcos): The implementation should be extended in the future to support
+        // multiple columns in `types` and to panic if the multi-column schema that ends up
+        // being too wide for a fixed-length N-byte pattern (MaterializeInc/materialize#22102).
+        assert!(N > 1);
+        assert!(types.len() == 1);
```

Comment:
I'd be surprised if the first assertion would make it into the compiled code at all. It is based on a static parameter `N` and can be checked at compile time.

The second assertion requires reading the length of a slice that is an argument to the function, so it is not an expensive computation? The documentation of `debug_assert` says the following: "The performance cost of assertions, however, is not measurable in general. Replacing [`assert!`] with `debug_assert!` is thus only encouraged after thorough profiling, and more importantly, only in safe code!"

I am delaying measurements to future PRs since it will be more feasible to run benchmark queries when we support more query constructs under type specialization (especially reductions).

Reply:
I tried to use `const_assert!`, but it has a problem with const generics, unfortunately: https://github.com/nvzqz/static-assertions/issues/40. I also tried the workaround mentioned in the issue, but it did not work differently than just having an `assert!` in there.